content/docs/io/async_read_write.md

```diff
@@ -1,12 +1,311 @@
+        pos: usize,
+    },
+    Empty,
+}
+
+// We want to be able to construct a ReadExact over anything
+// that implements AsyncRead, and any buffer that can be
+// thought of as a &mut [u8].
+fn read_exact<R, T>(reader: R, buffer: T) -> ReadExact<R, T>
+where
+    R: AsyncRead,
+    T: AsMut<[u8]>,
+{
+    ReadExact::Reading {
+        reader,
+        buffer,
+        // Initially, we've read no bytes into buffer.
+        pos: 0,
+    }
+}
+
+impl<R, T> Future for ReadExact<R, T>
+where
+    R: AsyncRead,
+    T: AsMut<[u8]>,
+{
+    // When we've filled up the buffer, we want to return both the buffer
+    // with the data that we read and the reader itself.
+    type Item = (R, T);
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        match *self {
+            ReadExact::Reading {
+                ref mut reader,
+                ref mut buffer,
+                ref mut pos,
+            } => {
+                let buffer = buffer.as_mut();
+                // Check that we haven't finished
+                while *pos < buffer.len() {
+                    // Try to read data into the remainder of the buffer.
+                    // Just like read in std::io::Read, poll_read *can* read
+                    // fewer bytes than the length of the buffer it is given,
+                    // and we need to handle that by looking at its return
+                    // value, which is the number of bytes actually read.
+                    //
+                    // Notice that we are using try_ready! here, so if poll_read
+                    // returns NotReady (or an error), we will do the same!
+                    // We uphold the contract that we have arranged to be
+                    // notified later because poll_read follows that same
+                    // contract, and _it_ returned NotReady.
+                    let n = try_ready!(reader.poll_read(&mut buffer[*pos..]));
+                    *pos += n;
+
+                    // If no bytes were read, but there was no error, this
+                    // generally implies that the reader will provide no more
+                    // data (for example, because the TCP connection was closed
+                    // by the other side).
+                    if n == 0 {
+                        return Err(io::Error::new(io::ErrorKind::UnexpectedEof, "early eof"));
+                    }
+                }
+            }
+            ReadExact::Empty => panic!("poll a ReadExact after it's done"),
+        }
+
+        // We need to return the reader and the buffer, which we can only
+        // do by moving them out of self. We do this by swapping our state
+        // with an "empty" one. This _should_ be fine, because poll() requires
+        // callers to not call poll() again after Ready has been returned,
+        // so we should only ever see ReadExact::Reading when poll() is called.
+        match mem::replace(self, ReadExact::Empty) {
```

Comment:
If we use an `Option<Reading>` instead of `Reading` be a variant of some `ReadExact` enum, we can avoid having to do the `mem::replace` and just use `Option::take`. Is that something we should consider?

Reply:
You'd need to wrap both the `buffer` and `reader` in `Option`, and then `take` them both. And you'd then also have to match on `Some` twice in `poll`. It would be possible, but not sure it's cleaner (except for avoiding the `mem::replace`). The reason I chose to do it this way was primarily because it matches the real implementation in `tokio`, but I could be convinced to change it :)

Reply:
I misunderstood you. Made the change in e06d56e07f4e0b76b34ca1a910c6453b8f978796!